include/socket_offload_ops.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2017 Texas Instruments
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Socket Offload Redirect API
+ */
+
+#ifndef __SOCKET_OFFLOAD_OPS_H__
+#define __SOCKET_OFFLOAD_OPS_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <net/net_ip.h>
+
+/*
+ * It is assumed that these offload functions (except for init()) follow the
+ * POSIX socket API standard for arguments, return values and setting of errno.
+ */
+struct socket_offload {
+	/* Perform one time initialization for the offload device: */
+	void (*init)();
+	/* POSIX Socket Functions: */
+	int (*socket)(int family, int type, int proto);
+	int (*close)(int sock);
+	int (*accept)(int sock, struct sockaddr *addr, socklen_t *addrlen);
+	int (*bind)(int sock, const struct sockaddr *addr, socklen_t addrlen);
+	int (*listen)(int sock, int backlog);
+	int (*connect)(int sock, const struct sockaddr *addr,
+		       socklen_t addrlen);
+	int (*setsockopt)(int sock, int level, int optname,
+			  const void *optval, socklen_t optlen);
+	int (*getsockopt)(int sock, int level, int optname, void *optval,
+			  socklen_t *optlen);
+	ssize_t (*recv)(int sock, void *buf, size_t max_len, int flags);
+	ssize_t (*recvfrom)(int sock, void *buf, short int len,
+			    short int flags, struct sockaddr *from,
+			    socklen_t *fromlen);
+	ssize_t (*send)(int sock, const void *buf, size_t len, int flags);
+	ssize_t (*sendto)(int sock, const void *buf, size_t len, int flags,
+			  const struct sockaddr *to, socklen_t tolen);
+};
+
+extern void socket_offload_register(const struct socket_offload *ops);
```

Comment:
Hmm, all the functions in socket_offload_ops.h have identical functionality in include/net/net_offload.h. Was there any reason include/net/net_offload.h could not be (re)used here or modified to fit this use case?

Reply:
Well, those functions require a net_context structure, and have callback arguments; and though they have similar names, the behaviour and usage is quite different.  The goal here is to provide an interface to BSD socket APIs provided by a vendors TCP/IP offload engine.